feat: use "prettier-tslint" and "prettier-eslint" when they exist

aleclarson opened the pull request:
When the scm root has `./node_modules/prettier-tslint` and/or `./node_modules/prettier-eslint`, we use them on any applicable modules.

We prefer `prettier-eslint` for `.js`, `.jsx`, and `.mjs` modules. If `prettier-eslint` does not exist, we try `prettier-tslint` and then `prettier` alone.

We prefer `prettier-tslint` for `.ts` and `.tsx`. If `prettier-tslint` does not exist, we try `prettier-eslint` and then `prettier` alone.

All other file types use `prettier` only.

aleclarson commented:
**PS:** You can add me as a maintainer here, too. ;)

azz commented on src/formatFiles.js:
```diff
@@ -1,27 +1,51 @@
+function loadFormat(rootDirectory, formatName, exportName) {
+  let entry;
+  try {
+    entry = require.resolve(join(rootDirectory, 'node_modules', formatName));
```
This won't work with yarn PnP.

aleclarson replied:
Are you saying it breaks `pretty-quick` for Yarn PnP installs, or just that `prettier-eslint` and `prettier-tslint` won't be found?

aleclarson replied:
Have you actually tested this with Yarn PnP?

azz replied:
The latter. Projects using PnP don't have `node_modules`

aleclarson replied:
Okay, that's fine. We can patch PnP support for this feature in a later patch.

aleclarson replied:
I've pushed a commit that resolves this.

azz commented:
Thinking I'd favour a more explicit API, like `pretty-quick --tslint`, `pretty-quick --eslint`. It will save a small amount of IO for users who don't use those tools, and users who have e.g. `prettier-eslint` still in `node_modules` (maybe unused) will be less confused when the behaviour changes after upgrading this package.

We can also make it work with PnP that way.

aleclarson commented:
> We can also make it work with PnP that way.

Why would replacing implicit loading with an explicit flag make PnP work? We already wrap the `require.resolve` call in a `try..catch` block, so PnP shouldn't choke on it.

azz commented:
We can just do `if (opts.eslint) require('prettier-tslint')`, shouldn't need to hardcode `node_modules`.

aleclarson commented:
Can you explain why someone would want `pretty-quick` to __not__ use `prettier-eslint` and/or `prettier-tslint` when they are installed?

aleclarson commented:
I've improved the implicit behavior by ensuring the resolved `prettier-tslint` and/or `prettier-eslint` paths always ignore globally-installed versions. Basically, they must be in your `package.json` to signal to `pretty-quick` that you're using them.

azz commented on src/formatFiles.js:
```diff
@@ -1,27 +1,61 @@
-    );
+
+    const format =
+      (/\.(mjs|jsx?)$/.test(file) && (eslintFormat || tslintFormat)) ||
```
This won't work for files with shebangs (which prettier supports). Maybe using `prettier.getFileInfo` would be better?

aleclarson replied:
Are you talking about executable files with no file extension that contain Javascript?

azz replied:
Absolutely.

azz replied:
https://github.com/prettier/prettier/pull/5149

aleclarson replied:
Let's support shebangs in a later patch. No biggie if they don't get linted for now.

aleclarson commented on src/formatFiles.js:
```diff
@@ -1,27 +1,61 @@
+function isGlobalModule(modulePath) {
+  const globalDirs = process.env.NODE_PATH.split(':');
+  for (let i = 0; i < globalDirs.length; i++) {
+    if (modulePath.startsWith(globalDirs[i] + '/')) return true;
```
Yeah, I can use `path.sep`

azz replied:
Be careful of cygwin/git bash which use `/` despite being on Windows.

aleclarson replied:
Are you saying the core `path` module does not account for that? See here: https://shapeshed.com/writing-cross-platform-node/#paths

aleclarson replied:
If I use the standard `path` module, that should be handled, I think. See here:
https://shapeshed.com/writing-cross-platform-node/#paths

aleclarson replied:
I've pushed a commit that resolves this.

aleclarson commented:
I've added shebang support (using `prettier.getFileInfo`), which requires Prettier v1.13.0 or higher.
I've updated the peer/dev dependencies on `prettier` to match.

Like I said, we can add Yarn PnP support in the future. I think we would take [this approach](https://github.com/arcanis/jest-pnp-resolver/blob/master/index.js).

LGTM! :)

aleclarson commented:
Hmm.. Looks like Travis is using a version of Prettier that doesn't include `getFileInfo`.

Even though I updated `yarn.lock`. 🤔 

@azz Could you take a look?

azz commented on package.json:
```diff
@@ -53,7 +53,10 @@
+    "prettier": "1.13.0",
     "semantic-release": "^11.0.2"
+  },
+  "jest": {
```
This should be fixed in newer versions of Jest, right?

aleclarson replied:
Are you asking if Jest 23.6.0 has this issue? Because it does.

azz commented on src/formatFiles.js:
```diff
@@ -1,27 +1,49 @@
+      format === prettierFormat
+        ? format(input, ((options.filepath = file), options))
+        : format({
+            text: input,
```
Could you add a test for this code path?

aleclarson replied:
I've tested locally (and you can too) with:

```sh
git clone -b feat1-test https://github.com/aleclarson/pretty-quick
```

azz commented on src/formatFiles.js:
```diff
@@ -1,27 +1,49 @@
+
+    const parser = getFileInfo.sync(file).inferredParser;
+    const format =
+      (parser === 'babylon' && (eslintFormat || tslintFormat)) ||
```
Or `flow`

aleclarson replied:
Does eslint support flow?

azz replied:
`babylon` supports flow language too. You can write JavaScript and still use the `flow` parser.

azz commented:
I deleted the build caches and restarted the build. Hopefully that helps.

aleclarson commented:
I think we can merge this.

Things to fix later:
- add Yarn PnP support
- get `prettier-eslint` to work with shebang modules

adrianolsk commented:
Hey guys, is this feature abandoned? :(

MichaelDeBoey commented:
Any news on this one? 🤔

Ping @aleclarson @azz

JounQin commented:
No thanks, but you can looking forward to https://github.com/prettier-eslint/prettier-plugin.
